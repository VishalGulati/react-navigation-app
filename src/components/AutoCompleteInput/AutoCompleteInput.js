import React, { Component } from 'react';
import PropTypes from 'prop-types';
import LocationsContext from '../../context/LocationsContext';

class AutoCompleteInput extends Component {
    constructor(props) {
        super(props);
        this.autocompleteInput = React.createRef();
        this.autocomplete = null;
    }

    static contextType = LocationsContext;

    handlePlaceChanged = () => {
        const place = this.autocomplete.getPlace();
        console.log(JSON.stringify(place.geometry.location));
        this.context.updateLocation(this.props.inputId, place.geometry.location)
    }

    componentDidUpdate() {
        if (this.context.mapLoaded) {
            this.autocomplete = new window.google.maps.places.Autocomplete(
                this.autocompleteInput.current,
                { types: ["geocode"] }
            );
            this.autocomplete.addListener("place_changed", this.handlePlaceChanged);
        }
    }

    render() {
        console.log('re-render')
        //const value = this.context[this.props.inputId];
        //console.log('value is : ', value)
        return (
            <input type="text"
                ref={this.autocompleteInput}
                className="form-control"
                id={this.props.inputId}
                placeholder="Enter a location" />
        );
    }
}

AutoCompleteInput.propTypes = {
    inputId: PropTypes.string,
    value: PropTypes.string
};

export default AutoCompleteInput;