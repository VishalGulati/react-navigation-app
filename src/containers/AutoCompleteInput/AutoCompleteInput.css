.autocomplete-input {
  width: 85%;
  display: inline-block;
}
.cross-icon {
  padding: 0 10px;
  color: black;
  font-weight: bold;
  font-size: 1.5rem;
  cursor: pointer;
  font-family: cursive;
}
