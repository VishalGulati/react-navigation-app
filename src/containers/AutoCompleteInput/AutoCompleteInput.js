import React, { Component } from 'react';
import PropTypes from 'prop-types';
import LocationsContext from '../../context/LocationsContext';
import './AutoCompleteInput.css';

/**
 * A re-usable component that returns a input field that is bound with Google maps API
 */
class AutoCompleteInput extends Component {
  constructor(props) {
    super(props);
    this.autocompleteInput = React.createRef();
    this.autocomplete = null;
    this.state = {
      isDirty: false
    };
  }

  static contextType = LocationsContext;

  handlePlaceChanged = () => {
    const place = this.autocomplete.getPlace();
    if (place && place.geometry) {
      this.context.updateLocation(this.props.inputId, place.geometry.location);
      this.setState({
        isDirty: true
      });
    }
  };

  resetField = () => {
    this.autocompleteInput.current.value = '';
    this.context.updateLocation(this.props.inputId, '');
    this.setState({
      isDirty: false
    });
  };

  componentDidUpdate() {
    if (this.context.mapLoaded) {
      this.autocomplete = new window.google.maps.places.Autocomplete(
        this.autocompleteInput.current,
        { types: ['geocode'] }
      );
      this.autocomplete.addListener('place_changed', this.handlePlaceChanged);
    }
    if (this.context.resetPending) {
      this.resetField();
    }
  }

  render() {
    return (
      <React.Fragment>
        <input
          type="text"
          ref={this.autocompleteInput}
          className="form-control autocomplete-input"
          id={this.props.inputId}
          placeholder="Enter a location"
        />
        {this.state.isDirty && (
          <span className="cross-icon" onClick={this.resetField}>
            X
          </span>
        )}
      </React.Fragment>
    );
  }
}

AutoCompleteInput.propTypes = {
  inputId: PropTypes.string
};

export default AutoCompleteInput;
