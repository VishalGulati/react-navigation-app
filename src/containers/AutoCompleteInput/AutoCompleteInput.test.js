import React from 'react';
import ReactDOM from 'react-dom';
import AutoCompleteInput from './AutoCompleteInput';
import { shallow } from 'enzyme';

const wrapper = shallow(<AutoCompleteInput />);

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AutoCompleteInput />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('Component has input element', () => {
    expect(wrapper.find('input').length).toEqual(1);
});


it('snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
});