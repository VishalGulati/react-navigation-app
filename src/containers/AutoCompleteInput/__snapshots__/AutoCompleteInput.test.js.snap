// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AutoCompleteInput />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <input
          className="form-control autocomplete-input"
          placeholder="Enter a location"
          type="text"
        />,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-control autocomplete-input",
          "id": undefined,
          "placeholder": "Enter a location",
          "type": "text",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <input
            className="form-control autocomplete-input"
            placeholder="Enter a location"
            type="text"
          />,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-control autocomplete-input",
            "id": undefined,
            "placeholder": "Enter a location",
            "type": "text",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
